packages/flutter/lib/src/material/stepper.dart

```diff
@@ -0,0 +1,648 @@
+/// steps. They are particularly useful in the case of forms where one step
+/// requires the completion of another one, or where multiple steps need to be
+/// completed in order to submit the whole form.
+///
+/// The widget is a flexible wrapper. A parent class should pass [currentStep]
+/// to this widget based on some logic triggered by the three callbacks that it
+/// provides.
+///
+/// See also:
+///
+///  * [Step]
+///  * <https://material.google.com/components/steppers.html>
+class Stepper extends StatefulWidget {
+  /// Creates a stepper from a list of steps.
+  ///
+  /// This widget is not meant to be rebuilt with a different set of steps. If
+  /// you need to swap the stepper, you will need to provide another key to it.
+  ///
+  /// The [steps], [axis], and [currentStep] arguments cannot be null.
+  Stepper({
+    Key key,
+    this.steps,
+    this.type: StepperType.vertical,
+    this.currentStep: 0,
+    this.onStepTapped,
+    this.onStepContinue,
+    this.onStepCancel
+  }) : super(key: key) {
+    assert(this.steps != null);
+    assert(this.type != null);
+    assert(this.currentStep != null);
+    assert(0 <= currentStep && currentStep < this.steps.length);
+  }
+
+  /// The steps of the stepper whose titles, subtitles, icons always get shown.
+  final List<Step> steps;
+
+  /// The type of stepper that determines the layout. In the case of
+  /// [StepperType.horizontal], the content of the current step is displayed
+  /// underneath as opposed to the [StepperType.vertical] case where it is
+  /// displayed in-between.
+  final StepperType type;
+
+  /// The index into [steps] of the current step whose content is displayed.
+  final int currentStep;
+
+  /// The callback called when a step is tapped, with its index passed as
+  /// an argument.
+  final ValueChanged<int> onStepTapped;
+
+  /// The callback called when the 'continue' button is tapped.
+  ///
+  /// If null, the 'continue' button will be disabled.
+  final VoidCallback onStepContinue;
+
+  /// The callback called when the 'cancel' button is tapped.
+  ///
+  /// If null, the 'cancel' button will be disabled.
+  final VoidCallback onStepCancel;
+
+  @override
+  _StepperState createState() => new _StepperState();
+}
+
+class _StepperState extends State<Stepper> {
+  List<GlobalKey> _keys;
+  final Map<int, StepState> _oldStates = new Map<int, StepState>();
+
+  @override
+  void initState() {
+    super.initState();
+    _keys = new List<GlobalKey>.generate(
+      config.steps.length,
+      (int i) => new GlobalKey()
+    );
+
+    for (int i = 0; i < config.steps.length; i += 1)
+      _oldStates[i] = config.steps[i].state;
+  }
+
+  @override
+  void didUpdateConfig(Stepper oldConfig) {
+    super.didUpdateConfig(oldConfig);
+    assert(config.steps.length == oldConfig.steps.length);
+
+    for (int i = 0; i < oldConfig.steps.length; i += 1)
+      _oldStates[i] = oldConfig.steps[i].state;
+  }
+
+  bool _first(int index) {
```

Comment:
this doesn't return the first boolean, so should probably be "_isFirst" or some such